### PR TITLE
Copy & Deployment Task changes for parallel execution and removal of alt...

### PR DESCRIPTION
**Pull request by kmkumaran**
Fixing alternate creds in Copy task
removing implicit copy from Deploy task
Adding parallel deploy support in Deploy task

### CODE REVIEW

File: Tasks/RunPowerShellOnTargetMachines/task.json [hunk 1]
```diff
@@ -16,6 +16,13 @@
         "Patch": 0,
         "IsTest": false
     },
+   "groups": [
```
Comment (pryenugu) [inline Tasks/RunPowerShellOnTargetMachines/task.json]: What does this group key do?

File: Tasks/RunPowerShellOnTargetMachines/task.json [hunk 2]
```diff
@@ -16,6 +16,13 @@
...
+        {
+            "name":"advanced",
+            "displayName":"Advanced",
+            "isExpanded":false
+        }
+    ],
     "inputs": [
         {
             "name": "EnvironmentName",
```
Comment (pryenugu) [inline Tasks/RunPowerShellOnTargetMachines/task.json]: Is the machine group label right? or does it need to be environment name only?

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 [hunk 3]
```diff
@@ -37,68 +37,17 @@ function Output-ResponseLogs
...
-}
-
 import-module "Microsoft.TeamFoundation.DistributedTask.Task.DevTestLabs"
 
-$env:DTL_ALTERNATE_CREDENTIALS_USERNAME = $alternateCredentialsUsername
-$env:DTL_ALTERNATE_CREDENTIALS_PASSWORD = $alternateCredentialsPassword
+$connection = Get-VssConnection -TaskContext $distributedTaskContext
 
-$resources = Get-EnvironmentResources -EnvironmentName $environmentName -ResourceFilter $machineNames -ErrorAction Stop
+$resources = Get-EnvironmentResources -EnvironmentName $environmentName -ResourceFilter $machineNames -ErrorAction Stop -Connection $connection
```
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: Have the -EA Stop in the end for all the cmdlets call here and the parameters to the cmdlets itself should be together.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 [hunk 4]
```diff
@@ -160,7 +109,7 @@ else
 	
         Write-Output "Copy will be Started for - $machine"
 
-        $job = Start-Job -ScriptBlock $CopyJob -ArgumentList $environmentName, $envOperationId, $machine, $sourcePath, $targetPath, $machineUserName, $machinePassword, $cleanTargetBeforeCopy
+        $job = Start-Job -ScriptBlock $CopyJob -ArgumentList $environmentName, $envOperationId, $machine, $sourcePath, $targetPath, $machineUserName, $machinePassword, $cleanTargetBeforeCopy, $connection
 
         $Jobs.Add($job.Id, $machine)
```
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: Have the formatting right in "Copy Starter for    - $machine". To many whitespaces.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 [hunk 5]
```diff
@@ -110,7 +59,7 @@ if($deployFilesInParallel -eq "false")
         $fdqn = $resource.Name
         Write-Output "Copy will be Started for - $fdqn"
```
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: Write-Output and write-host both appear to the console in the build definition summary console view?

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 [hunk 6]
```diff
@@ -110,7 +59,7 @@ if($deployFilesInParallel -eq "false")
         $fdqn = $resource.Name
```
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: is there supposed to be an FQDN property bag? or we will implement that later to pick it up from?

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 4 above)
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: we are making two calls to Get-Job in the while loop and then immediately in the foreach loop after 10 seconds. Can we look to see how to optimize that?

Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: I would recommend to have the Remove-job right after Recieve-Jobm in that way later if any code in between goes awry, we know we are not stuck in an infinite loop. Just some good coding practice.

Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: aren't we getting the machineName in the output (deploymentResponse)? if that is the case should we open a bug at the DTL SDK layer? We are maintaining a hash table here with job ID's and machineNames which seems wrong to me.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 [hunk 7]
```diff
@@ -4,9 +4,7 @@ param (
     [string]$sourcePath,
```
Comment (pryenugu) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: Should we start taking port as well in this task? We are hardcoding the port right now. Maybe open as a bug?

File: Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1 [hunk 8]
```diff
@@ -2,21 +2,22 @@ param (
...
 Write-Verbose "machineNames = $machineNames" -Verbose
 Write-Verbose "sourcePackage = $sourcePackage" -Verbose
-Write-Verbose "applicationPath = $applicationPath" -Verbose
 Write-Verbose "scriptPath = $scriptPath" -Verbose
 Write-Verbose "scriptArguments = $scriptArguments" -Verbose
 Write-Verbose "initializationScriptPath = $initializationScriptPath" -Verbose
 
+. ./RunPowerShellJob.ps1
+
 function Output-ResponseLogs
```
Comment (pryenugu) [inline Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1]: We should have this function in a common module, as we are duplicating this now in many tasks. Just doesn't seem right to me.

File: Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1 [hunk 9]
```diff
@@ -55,50 +56,86 @@ $credential = New-Object 'System.Net.NetworkCredential' -ArgumentList $machineUs
...
+
+                 if($status -ne "Passed")
+                 {
+                     $envOperationStatus = "Failed"
+                 }
+            
+                 $machineName = $Jobs.Item($job.Id)
+
+                 Output-ResponseLogs -operationName "Deployment" -fqdn $machineName -deploymentResponse $output
+                 Write-Output "Result for machine $machineName : $result"
```
Comment (pryenugu) [inline Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1]: We should be removing these write-output's which are just printing duplicate information which are already in the write-verbose in Output-ResponseLogs. In my opinion it is cluttering our logs. We should not print this on console, as it is not live and does not make sense. Just printing "Completed" will suffice. Any thoughts?

File: Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1 [hunk 10]
```diff
@@ -55,50 +56,86 @@ $credential = New-Object 'System.Net.NetworkCredential' -ArgumentList $machineUs
...
+        Write-Output "Deployment will be Started for - $machine"
 
-        throw $response.Error;
+        $job = Start-Job -ScriptBlock $RunPowershellJob -ArgumentList $environmentName, $envOperationId, $machine, $scriptPath, $port, $scriptArguments, $initializationScriptPath, $credential, $connection
+
+        $Jobs.Add($job.Id, $machine)
+    
+        Write-Output "Deployment Started for  - $machine"
+    }
+    While (Get-Job)
```
Comment (pryenugu) [inline Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1]: I can see that the logic of this get-job is also shared, and it should be put in a shared module as well.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 7 above)
Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: we have a separate story for that

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 6 above)
Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: that we will handle in a separate story.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 5 above)
Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: Write-Output will show in console

File: Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1 (hunk 9 above)
Comment (niadak) [inline Tasks/RunPowerShellOnTargetMachines/RunPowerShellOnTargetMachines.ps1]: lets remove the log one. Printing deployment status is ok I believe.

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 4 above)
Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: extra space I will remove.

Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: the logic of calling Get-Job twice is, first call check if there is any job is there and if so we give 10 sec time to complete that job and do a Get-Job again. Then check the Job status for the jobs.

Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: moving Remove-Job right after receive-Job seems right to me. I will do that

File: Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1 (hunk 3 above)
Comment (niadak) [inline Tasks/CopyFilesToMachines/CopyFilesToMachines.ps1]: sure. will do that

**Comment by kmkumaran** [comment]
Addressed all review comments. Closing the review.

**Pull request merged**